rpc_util.go

```diff
@@ -635,61 +640,61 @@ type payloadInfo struct {
 	uncompressedBytes []byte
 }
 
-func recvAndDecompress(p *parser, s *transport.Stream, dc Decompressor, maxReceiveMessageSize int, payInfo *payloadInfo, compressor encoding.Compressor) ([]byte, error) {
-	pf, d, err := p.recvMsg(maxReceiveMessageSize)
+func recvAndDecompress(p *parser, s *transport.Stream, dc Decompressor, maxReceiveMessageSize int, payInfo *payloadInfo, compressor encoding.Compressor) error {
+	pf, err := p.recvMsg(maxReceiveMessageSize)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	if payInfo != nil {
-		payInfo.wireLength = len(d)
+		payInfo.wireLength = len(p.msg)
 	}
 
 	if st := checkRecvPayload(pf, s.RecvCompress(), compressor != nil || dc != nil); st != nil {
-		return nil, st.Err()
+		return st.Err()
 	}
 
 	if pf == compressionMade {
 		// To match legacy behavior, if the decompressor is set by WithDecompressor or RPCDecompressor,
 		// use this decompressor as the default.
 		if dc != nil {
-			d, err = dc.Do(bytes.NewReader(d))
+			p.msg, err = dc.Do(bytes.NewReader(p.msg))
 			if err != nil {
-				return nil, status.Errorf(codes.Internal, "grpc: failed to decompress the received message %v", err)
+				return status.Errorf(codes.Internal, "grpc: failed to decompress the received message %v", err)
 			}
 		} else {
-			dcReader, err := compressor.Decompress(bytes.NewReader(d))
+			dcReader, err := compressor.Decompress(bytes.NewReader(p.msg))
 			if err != nil {
-				return nil, status.Errorf(codes.Internal, "grpc: failed to decompress the received message %v", err)
+				return status.Errorf(codes.Internal, "grpc: failed to decompress the received message %v", err)
 			}
 			// Read from LimitReader with limit max+1. So if the underlying
 			// reader is over limit, the result will be bigger than max.
-			d, err = ioutil.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1))
+			p.msg, err = ioutil.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1))
```

Comment:
In this line ioutil.ReadAll allocates a new buffer and then p.msg gets thrown away. This is still an improvement because a compressed message results in one buffer allocation instead of two, but not ideal.

It might be best to use a `sync.Pool` to store the buffers.

Reply:
yeah, that's true. we don't use compression in our applications so I haven't bothered trying to optimize those codepaths :)

Reply:
If you don't intend to optimize this code path, I think it would be better to not overwrite `p.msg` here. Assign the result of `ReadAll` to a new variable. That way `p.msg` can be reused for receiving bytes on the wire messages, a new buffer will be allocated every time we need to decompress.